fastlane/swift/Runner.swift

```diff
@@ -22,14 +22,47 @@ func desc(_: String) {
     // no-op, this is handled in fastlane/lane_list.rb
 }
 
+class AtomicDictionary<Key: Hashable, Value> {
+    private let lockQueue = DispatchQueue(label: "tools.fastlane.serial.queue")
+    private var storage: [Key: Value]
+
+    init() {
+        storage = [:]
+    }
+
+    func get(_ key: Key) -> Value? {
+        lockQueue.sync {
+            storage[key]
+        }
+    }
+
+    func set(_ key: Key, value: Value) {
+        lockQueue.sync {
+            storage[key] = value
+        }
+    }
+
+    func removeValue(forKey key: Key) {
+        lockQueue.sync {
+            _ = storage.removeValue(forKey: key)
+        }
+    }
+
+    var allValues: [Key: Value] {
+        lockQueue.sync {
+            storage
+        }
+    }
+}
+
 class Runner {
     private var thread: Thread!
     private var socketClient: SocketClient!
     private let dispatchGroup = DispatchGroup()
     private var returnValue: String? // lol, so safe
     private var currentlyExecutingCommand: RubyCommandable?
     private var shouldLeaveDispatchGroupDuringDisconnect = false
-    private var executeNext: [String: Bool] = [:]
+    private var executeNext: AtomicDictionary<String, Bool> = AtomicDictionary()
```

Comment:
there shouldn't be any need of an atomic usage here, because all reads and writes of the dictionary are made inside a `DispatchGroup`, so no two callers can access to that method at once, that way we preserve the lock and the order of the operations.

Reply:
There was a crash with that code, check my comments previous to adding this part of the fix: https://github.com/fastlane/fastlane/issues/18502#issuecomment-949980464

Reply:
I read that crash but I don't see how it is related to the `DispatchGroup` being broken. Not that the introduced code is so breaking, but I just don't see any relationship. Again, I don't have a M1 machine so I can't see the traceback from Xcode about that method crashing.

Reply:
I've unapplied AtomicDictionary commit to get a screenshot with the crash and the trace by the ThreadSanitizer. Just in case it helps any better.
![Screenshot 2021-12-01 at 15 28 50](https://user-images.githubusercontent.com/2754671/144252592-fb3500bd-95d4-4681-aa6c-5ea2855d5076.png)

Reply:
We should figure out why there're being two concurrent accesses to that method in first place I think it might be the root of the problem🤔

Reply:
If I may, are you thinking of `DispatchSemaphore` instead of `DispatchGroup`? I am not sure how `DispatchGroup` protects against concurrent access here (entering a group is not a wait operation).

Reply:
I don't think that it was added as a exclusion zone but more for a balancing calls method (I'm not the author of the whole implementation, I'm just the maintainer but I don't have an M1 to help here), maybe this doesn't do anything at all, but if iirc, this was needed for synchronizing with other parts of the code.

Reply:
Here it's the reason of the data race:

- waitWithPolling is retaining/reading  executeNext in one thread
- commandExecuted is writing executeNext in other thread

<img width="679" alt="Screenshot 2021-12-02 at 01 19 07" src="https://user-images.githubusercontent.com/2754671/144336124-23688243-ba4c-4e56-9721-89636577d6ce.png">

<img width="2035" alt="Screenshot 2021-12-02 at 01 10 51" src="https://user-images.githubusercontent.com/2754671/144336421-fc019b1e-6a45-4374-8429-78620cda325f.png">

Reply:
To not block the current running thread (so you can poll it) the wait is being performed in another thread, yes, that's correct.

Reply:
```suggestion
    private var executeNext: AtomicDictionary<String, Bool> = .init()
```